13_BitonicSort/main.cpp

```diff
@@ -0,0 +1,323 @@
+
+		const nbl::asset::SPushConstantRange pcRange = { .stageFlags = IShader::E_SHADER_STAGE::ESS_COMPUTE,.offset = 0,.size = sizeof(BitonicPushData) };
+
+		smart_refctd_ptr<IGPUPipelineLayout> layout;
+		smart_refctd_ptr<IGPUComputePipeline> bitonicSortPipeline;
+		{
+			layout = m_device->createPipelineLayout({ &pcRange,1 });
+			IGPUComputePipeline::SCreationParams params = {};
+			params.layout = layout.get();
+			params.shader.shader = bitonicSortShader.get();
+			params.shader.entryPoint = "main";
+			params.shader.entries = nullptr;
+			if (!m_device->createComputePipelines(nullptr, { &params,1 }, &bitonicSortPipeline))
+				return logFail("Failed to create compute pipeline!\n");
+		}
+
+		nbl::video::IDeviceMemoryAllocator::SAllocation allocation[4] = {};
+		smart_refctd_ptr<IGPUBuffer> buffers[4];
+
+		auto build_buffer = [this](
+			smart_refctd_ptr<ILogicalDevice> m_device,
+			nbl::video::IDeviceMemoryAllocator::SAllocation* allocation,
+			smart_refctd_ptr<IGPUBuffer>& buffer,
+			size_t buffer_size,
+			const char* label
+			) -> void {
+				IGPUBuffer::SCreationParams params;
+				params.size = buffer_size;
+				params.usage = IGPUBuffer::EUF_STORAGE_BUFFER_BIT | IGPUBuffer::EUF_SHADER_DEVICE_ADDRESS_BIT;
+				buffer = m_device->createBuffer(std::move(params));
+				if (!buffer)
+					logFail("Failed to create GPU buffer of size %d!\n", buffer_size);
+
+				buffer->setObjectDebugName(label);
+
+				auto reqs = buffer->getMemoryReqs();
+				reqs.memoryTypeBits &= m_physicalDevice->getHostVisibleMemoryTypeBits();
+
+				*allocation = m_device->allocate(reqs, buffer.get(), IDeviceMemoryAllocation::EMAF_DEVICE_ADDRESS_BIT);
+				if (!allocation->isValid())
+					logFail("Failed to allocate Device Memory compatible with our GPU Buffer!\n");
+
+				assert(allocation->memory.get() == buffer->getBoundMemory().memory);
+			};
+
+		build_buffer(m_device, allocation, buffers[0], sizeof(uint32_t) * element_count, "Input Key Buffer");
+		build_buffer(m_device, allocation + 1, buffers[1], sizeof(uint32_t) * element_count, "Input Value Buffer");
+		build_buffer(m_device, allocation + 2, buffers[2], sizeof(uint32_t) * element_count, "Output Key Buffer");
+		build_buffer(m_device, allocation + 3, buffers[3], sizeof(uint32_t) * element_count, "Output Value Buffer");
+
+		uint64_t buffer_device_address[] = {
+			buffers[0]->getDeviceAddress(),
+			buffers[1]->getDeviceAddress(),
+			buffers[2]->getDeviceAddress(),
+			buffers[3]->getDeviceAddress()
+		};
+
+	
+		void* mapped_memory[] = {
+			allocation[0].memory->map({0ull,allocation[0].memory->getAllocationSize()}, IDeviceMemoryAllocation::EMCAF_READ),
+			allocation[1].memory->map({0ull,allocation[1].memory->getAllocationSize()}, IDeviceMemoryAllocation::EMCAF_READ),
+			allocation[2].memory->map({0ull,allocation[2].memory->getAllocationSize()}, IDeviceMemoryAllocation::EMCAF_READ),
+			allocation[3].memory->map({0ull,allocation[3].memory->getAllocationSize()}, IDeviceMemoryAllocation::EMCAF_READ)
```

Comment:
Vulkan doesn't care about the EMCAF flags, but they're kinda wrong here, I'm pretty sure you WRITE into them with the host, while others you READ at the end